### PR TITLE
Draft/xsimd

### PR DESCRIPTION
**Pull request by serge-sans-paille**

### CODE REVIEW

File: content/2018-10-31-xsimd.rst [hunk 1]
```diff
@@ -0,0 +1,271 @@
...
+          motivation of this change, and the advantage of this move.
+
+`boost.simd <https://github.com/NumScale/boost.simd>`_ provides a C++
+abstraction of vector type, allowing for efficient vectorization of array
+computations. It has been (optionally) used as part of the expression template
+engine of Pythran for a long time, a great collaboration that led to several
+patch in boost.simd, and great performance for Pythran.
+
+Unfortunately, the project has been silent over the last months (see for
+instance `this issue <https://github.com/NumScale/boost.simd/issues/546>`_ and
```
Comment (HadrienG2) [inline content/2018-10-31-xsimd.rst]: You should close that parenthesis.

File: content/2018-10-31-xsimd.rst [hunk 2]
```diff
@@ -0,0 +1,271 @@
...
+    clang + boost.simd: 220 us +- 8 us
+    clang + xsimd:      273 us +- 11 us
+    clang:             1.41 ms +- 0.04 ms
+
+Interestingly, **on that example**, Clang generates better code for the boost.simd version. So let's be wary of hasty conclusion and just state that with both engines, I can get efficient vectorization of my code.
+
+Complex Numbers
+===============
+
+Thanks to xsimd, Pythran is no able to *naively* support complex number
```
Comment (HadrienG2) [inline content/2018-10-31-xsimd.rst]: now*

File: content/2018-10-31-xsimd.rst [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+
+
+With xsimd, it's slightly faster (no boost dependencies, and less C++ magic).
+
+.. code-block:: sh
+
+    time pythran -O3 -march=native arc_distance.cpp -DUSE_XSIMD
+    10.32s user 0.65s system 102% cpu 10.688 total
+
+Arguably, it's still slow. But a bit less then what it uses to be, and the reason also lies in the generated c++, not only in the vectorization engine:
```
Comment (HadrienG2) [inline content/2018-10-31-xsimd.rst]: This explanation is a bit hard to follow. I think grouping the timing of the initial cpp file generation and boost.simd parts at the beginning does more harm than good on the clarity front. It could be easier to understand in this order/grouping:

- Generation of cpp and raw pythran compile time ("reference")
- Boost.SIMD version ("before")
- xsimd version ("after")

File: content/2018-10-31-xsimd.rst [hunk 4]
```diff
@@ -0,0 +1,271 @@
...
+worse than boost.simd implementation*. But maybe there are bad interactions
+with Pythran's expression template? A single register spill could wreak havoc
+in the performance, and I know there is room for improvement there.
+
+Conclusions
+===========
+
+I'm indeed very happy with the changes. The xsimd team is very reactive, it's
+cool to chat with them about performance, Python, C++... And did I see xsimd
+supports Neon, avx512? I should try to run cross-compiled Pythran code on a
```
Comment (HadrienG2) [inline content/2018-10-31-xsimd.rst]: Upstream capitalization is NEON and AVX512 (not sure why for NEON actually, since that is never spelled out as an acronym anywhere in ARM's documentation, but upstream is always right :smiley: )

File: content/2018-10-31-xsimd.rst [hunk 5]
```diff
@@ -0,0 +1,271 @@
...
+Raspberry, but... That's for another story!
+
+Again thanks a lot to (alphabetical order) `Johan
+<https://twitter.com/JohanMabille>`_, `martin
```
Comment (HadrienG2) [inline content/2018-10-31-xsimd.rst]: Martin?

File: content/2018-10-31-xsimd.rst [hunk 6]
```diff
@@ -0,0 +1,271 @@
...
+<https://twitter.com/renou_martin>`_, Sylvain
```
Comment (HadrienG2) [inline content/2018-10-31-xsimd.rst]: Missing backtick before "Sylvain" breaks the link.

File: content/2018-10-31-xsimd.rst [hunk 7]
```diff
@@ -0,0 +1,271 @@
...
+computations. It has been (optionally) used as part of the expression template
+engine of Pythran for a long time, a great collaboration that led to several
+patch in boost.simd, and great performance for Pythran.
+
+Unfortunately, the project has been silent over the last months (see for
+instance `this issue <https://github.com/NumScale/boost.simd/issues/546>`_ and
+I had to maintain a few custom patches. Turns out the project has been
+re-branded as **bSIMD** with a more restrictive license, as detailed in `another
+issue <https://github.com/NumScale/boost.simd/issues/545>`_. From the Pythran
+perspective, this is no good news.
```
Comment (HadrienG2) [inline content/2018-10-31-xsimd.rst:23]: Technically, they have recently started talking about a freeware fork again (see the bottom of the linked issue). But practically, the way they have (mis)handled this killed my confidence in this project and made me question whether the NumScale economic model is sound at all. I guess same goes for you.

You may or may not want to expand on this.

File: content/2018-10-31-xsimd.rst [hunk 8]
```diff
@@ -0,0 +1,271 @@
...
+
+Fortunately, the people from `QuantStack <http://quantstack.net/>`_ have put
+tremendous effort into providing an equivalent to boost.simd, *xsimd*. And
```
Comment (HadrienG2) [inline content/2018-10-31-xsimd.rst]: Direct link to http://quantstack.net/xsimd.html?

File: content/2018-10-31-xsimd.rst [hunk 9]
```diff
@@ -0,0 +1,271 @@
...
+their library actually provides some improvements in the context of Pythran, so
+when they proposed to fund the move to *xsimd*, it was just perfect.
+
+So here is the deal: I do the port, report any API and/or performance issue,
+and eventually provide patch when relevant. That's what I did over the last
```
Comment (HadrienG2) [inline content/2018-10-31-xsimd.rst]: patches*

File: content/2018-10-31-xsimd.rst [hunk 10]
```diff
@@ -0,0 +1,271 @@
...
+three months, let's have a look at the result.
```
Comment (HadrienG2) [inline content/2018-10-31-xsimd.rst]: results?

File: content/2018-10-31-xsimd.rst [hunk 11]
```diff
@@ -0,0 +1,271 @@
...
+
+User-level Changes
+==================
+
+In order to activate explicit vectorisation, one must pass ``-DUSE_XSIMD -march=native`` to the pythran compiler, in place of ``-DUSE_BOOST_SIMD -march=native``. Fair enough.
```
Comment (HadrienG2) [inline content/2018-10-31-xsimd.rst]: s/pythran/Pythran

(Pardon my capitalism)

File: content/2018-10-31-xsimd.rst [hunk 12]
```diff
@@ -0,0 +1,271 @@
...
+
+For instance, consider the following kernel, taken from the `numpy benchmarks
+suite <https://github.com/serge-sans-paille/numpy-benchmarks/>`_ suite I
```
Comment (HadrienG2) [inline content/2018-10-31-xsimd.rst]: suite suite?

File: content/2018-10-31-xsimd.rst [hunk 13]
```diff
@@ -0,0 +1,271 @@
...
+
+Compilation Time
+================
+
+Pythran is an *Ahead of Time* compiler, so compilation time is generally not a
+good metric. But there's one situation where it matters to me: Continuous
+Integration. Because Travis has time limits, the faster we compile, the more
+test we can pass! As Pythran validates for Python2 and Python3, for Clang and
+GCC, with and without SIMD, with and without OpenMP, that's a lot of
+configuration to test. Roughly... 20hours of cumulated tests actually, see
```
Comment (HadrienG2) [inline content/2018-10-31-xsimd.rst]: configurations?

File: content/2018-10-31-xsimd.rst [hunk 14]
```diff
@@ -0,0 +1,271 @@
...
+        10.05μs     9.56μs      0.95  benchmarks.TimeSuite.time_square_array
+    -   56.82μs    45.32μs      0.80  benchmarks.TimeSuite.time_tan_array
+    -   98.85μs    44.16μs      0.45  benchmarks.TimeSuite.time_tanh_array
+
+Wow, that's significant changes. Regression on both ``cos``, ``sin`` and ``acos`` are not good news.
+
+What conclusion should we draw? My take on this is that these benchmarks are
+not synthetic enough to state *xsimd implementation of function X is better or
+worse than boost.simd implementation*. But maybe there are bad interactions
+with Pythran's expression template? A single register spill could wreak havoc
```
Comment (HadrienG2) [inline content/2018-10-31-xsimd.rst]: templates?

File: content/2018-10-31-xsimd.rst [hunk 15]
```diff
@@ -0,0 +1,271 @@
...
+in the performance, and I know there is room for improvement there.
+
+Conclusions
+===========
+
+I'm indeed very happy with the changes. The xsimd team is very reactive, it's
+cool to chat with them about performance, Python, C++... And did I see xsimd
```
Comment (HadrienG2) [inline content/2018-10-31-xsimd.rst]: say?

**Review by HadrienG2 — CHANGES_REQUESTED** [review]

File: content/2018-10-31-xsimd.rst (hunk 11 above)
Comment (HadrienG2) [inline content/2018-10-31-xsimd.rst]: Also, vectorization*

File: content/2018-10-31-xsimd.rst [hunk 16]
```diff
@@ -0,0 +1,271 @@
...
+write. Good point.
+
+
+Compilation Time
+================
+
+Pythran is an *Ahead of Time* compiler, so compilation time is generally not a
+good metric. But there's one situation where it matters to me: Continuous
+Integration. Because Travis has time limits, the faster we compile, the more
+test we can pass! As Pythran validates for Python2 and Python3, for Clang and
```
Comment (HadrienG2) [inline content/2018-10-31-xsimd.rst]: tests?

File: content/2018-10-31-xsimd.rst [hunk 17]
```diff
@@ -0,0 +1,272 @@
...
+:authors: serge-sans-paille
+:summary: After five years of friendship, Pythran replaces ``boost.simd`` by
+          ``xsimd`` as its vectorization engine. This post goes through the
+          motivation of this change, and the advantage of this move.
+
+`boost.simd <https://github.com/NumScale/boost.simd>`_ provides a C++
+abstraction of vector type, allowing for efficient vectorization of array
+computations. It has been (optionally) used as part of the expression template
+engine of Pythran for a long time, a great collaboration that led to several
+patch in boost.simd, and great performance for Pythran.
```
Comment (JohanMabille) [inline content/2018-10-31-xsimd.rst]: patch => patches

File: content/2018-10-31-xsimd.rst [hunk 18]
```diff
@@ -0,0 +1,272 @@
...
+And with vectorization on (note that the calls to ``np.min()`` do not involve complex numbers though.
+
+.. code-block:: sh
+
+    CC=gcc CXX=g++ pythran -march=native -O3 make_decision.py -DUSE_XSIMD
+    python -m perf timeit -s 'import numpy as np; np.random.seed(0); N = 100000; x = np.random.random(N) + 1j *  np.random.random(N); from normalize_complex_arr import normalize_complex_arr' 'normalize_complex_arr(x)'
+    .....................
+    Mean +- std dev: 723 us +- 14 us
+
+Cool! Speedup for complex! For the record, the numpy version already ran at roughly ``3.19 ms +- 0.02 ms``.
```
Comment (JohanMabille) [inline content/2018-10-31-xsimd.rst:142]: I will put that result first to keep the narrative order numpy => pythran => pythran + xsimd

File: content/2018-10-31-xsimd.rst [hunk 19]
```diff
@@ -0,0 +1,272 @@
...
+write. Good point.
+
+
+Compilation Time
+================
+
+Pythran is an *Ahead of Time* compiler, so compilation time is generally not a
+good metric. But there's one situation where it matters to me: Continuous
+Integration. Because Travis has time limits, the faster we compile, the more
+test we can pass! As Pythran validates for Python2 and Python3, for Clang and
```
Comment (JohanMabille) [inline content/2018-10-31-xsimd.rst]: test => tests

File: content/2018-10-31-xsimd.rst [hunk 20]
```diff
@@ -0,0 +1,272 @@
...
+    .....................
+    Mean +- std dev: 723 us +- 14 us
+
+Cool! Speedup for complex! For the record, the numpy version already ran at roughly ``3.19 ms +- 0.02 ms``.
+
+Scalar Version
+==============
+
+That's probably a detail for many xsimd users, but thanks to this cooperation,
+xsimd now exposes a scalar version of all its mathematical function inside the
```
Comment (JohanMabille) [inline content/2018-10-31-xsimd.rst]: all its mathematical function => all the mathematical functions

File: content/2018-10-31-xsimd.rst [hunk 21]
```diff
@@ -0,0 +1,272 @@
...
+
+.. code-block:: sh
+
+    CC=gcc CXX=g++ pythran -march=native -O3 normalize_complex_arr.py
+    python -m perf timeit -s 'import numpy as np; np.random.seed(0); N = 100000; x = np.random.random(N) + 1j *  np.random.random(N); from normalize_complex_arr import normalize_complex_arr' 'normalize_complex_arr(x)'
+    .....................
+    Mean +- std dev: 2.84 ms +- 0.01 ms
+
+
+And with vectorization on (note that the calls to ``np.min()`` do not involve complex numbers though.
```
Comment (JohanMabille) [inline content/2018-10-31-xsimd.rst]: I find it a bit confusing because you want to show speed up thanks to complex vectorization but you mention a non complex function. I think stating that "with vectorization enabled, not only the real parts of the ``normalize_complex_arr`` function are accelerated, but the whole function is", could be clearer; there might be a better formulation than mine, but the idea is to point out which parts of the function are now accelerated thanks to complex vectorisation support.

File: content/2018-10-31-xsimd.rst [hunk 22]
```diff
@@ -0,0 +1,282 @@
...
+when they proposed to fund the move to *xsimd*, it was just perfect.
+
+So here is the deal: I do the port, report any API and/or performance issue,
+and eventually provide patches when relevant. That's what I did over the last
+three months, let's have a look at the results.
+
+User-level Changes
+==================
+
+In order to activate explicit vectorisation, one must pass ``-DUSE_XSIMD -march=native`` to the Pythran compiler, in place of ``-DUSE_BOOST_SIMD -march=native``. Fair enough.
```
Comment (wolfv) [inline content/2018-10-31-xsimd.rst:38]: This might be a bit off-topic, but why not change the SIMD flag to something generic, like `-DUSE_EXPLICIT_SIMD`
